lib/spack/spack/compilers/__init__.py

```diff
@@ -188,19 +189,53 @@ def all_compiler_specs(scope=None, init_config=True):
 
 
 def find_compilers(path_hints=None):
-    """Returns the list of compilers found in the paths given as arguments.
+    """Returns the list of compilers found in the paths or modules given.
+
+    If a compiler is found by both path and module, use the module version.
+    Note: all compiler detection for Cray systems is done by modules. Module
+    hints may not be provided for Cray systems.
 
     Args:
-        path_hints (list or None): list of path hints where to look for.
-            A sensible default based on the ``PATH`` environment variable
-            will be used if the value is None
+        path_hints (list or None): list of path hints in which to look for
+            compilers. A sensible default based on the ``PATH`` environment
+            variable will be used if the value is None.
+
+        module_hints (list or None): list of modules in which to look for
+            compilers. A sensible default based on ``module avail`` will be
+            used if the values is None.
 
     Returns:
         List of compilers found
     """
+    # Get modules if none given
+    if spack.architecture.platform().name == "cray":
+        # Cray modules are handled in their own special way
+        modules_to_check_paths = []
+    else:
+        avail_output = spack.util.module_cmd.module('avail')
+        modules_to_check_paths = avail_output.split()
+
+    # Associate paths with modules. If a compiler is found in the path
+    # associated with a module, it will include that module
+    module_association = {}
+    for mod in modules_to_check_paths:
+        path = spack.util.module_cmd.get_bin_path_from_module(mod)
```

Comment:
I don't think it will end up with `icc` modules being associated with the `gcc` compilers. `icc` modules should be adding `icc` to the `PATH` and both `icc` and `gcc` to the `LD_LIBRARY_PATH`. That's precisely why I'm not using the existing functionality to get a path from a module, because it would check `LD_LIBRARY_PATH` as well.